File: R/Endpoint Examples/Multipart Payload/markdown.R
#!
# What this sample does:
# - Converts a PDF to Markdown using pdfRest.
# - Sends a single multipart/form-data request directly to /markdown with the file.
#
# Setup (.Renviron):
# - Copy .Renviron.example to .Renviron (R folder root)
# - Set PDFREST_API_KEY=your_api_key_here
# - Optional: set PDFREST_URL to override the API region. For EU/GDPR compliance and proximity, use:
#     PDFREST_URL=https://eu-api.pdfrest.com
#
# Usage:
#   Rscript "Endpoint Examples/Multipart Payload/markdown.R" /path/to/input.pdf
#
# Output:
# - Prints the API JSON response to stdout. Non-2xx responses quit with a concise message.
# - Tip: pipe output to a file: Rscript ... > response.json

suppressWarnings(suppressMessages({
  if (!requireNamespace("httr", quietly = TRUE)) stop("Please install 'httr' package")
}))

stderrf <- function(...) cat(sprintf(...), file = stderr())

api_key <- Sys.getenv("PDFREST_API_KEY", unset = "")
if (identical(api_key, "")) {
  stderrf("Missing PDFREST_API_KEY in environment (.Renviron or shell)\n")
  quit(status = 1)
}

api_base <- sub("/+$$", "", Sys.getenv("PDFREST_URL", unset = "https://api.pdfrest.com"))

args <- commandArgs(trailingOnly = TRUE)
pdf_path <- args[1]
if (is.null(pdf_path) || !file.exists(pdf_path)) {
  stderrf("Usage: Rscript markdown.R /path/to/file.pdf\n")
  quit(status = 1)
}

filename <- basename(pdf_path)

tryCatch({
  conn_url <- paste0(api_base, "/markdown")

  # Build multipart form body. httr sets multipart/form-data with boundary.
  body <- list(
    file = httr::upload_file(pdf_path, type = "application/pdf"),
    output = "pdfrest_markdown",
    page_break_comments = "on"
    # Optional parameters:
    # page_range = "1-3"
  )

  resp <- httr::POST(
    conn_url,
    httr::add_headers("api-key" = api_key),
    body = body,
    encode = "multipart"
  )

  txt <- httr::content(resp, as = "text", encoding = "UTF-8")
  cat(txt)
  if (httr::http_error(resp)) {
    stop(sprintf("Markdown conversion failed with status %s", httr::status_code(resp)))
  }

}, error = function(e) {
  stderrf("Error: %s: %s\n", class(e)[1], conditionMessage(e))
  quit(status = 1)
})

File: Perl/Endpoint Examples/Multipart Payload/markdown.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
use FindBin qw($Bin);
use File::Basename qw(basename);
use LWP::UserAgent;
use HTTP::Request::Common qw(POST);
use Dotenv;

#!
# What this sample does:
# - Converts a PDF to Markdown using pdfRest.
# - Sends a single multipart/form-data request directly to /markdown with the file.
#
# Setup (.env):
# - Copy .env.example to .env (Perl folder root)
# - Set PDFREST_API_KEY=your_api_key_here
# - Optional: set PDFREST_URL to override the API region. For EU/GDPR compliance and proximity, use:
#     PDFREST_URL=https://eu-api.pdfrest.com
# For more information visit https://pdfrest.com/pricing#how-do-eu-gdpr-api-calls-work
#
# Usage:
#   perl "Endpoint Examples/Multipart Payload/markdown.pl" /path/to/input.pdf
#
# Output:
# - Prints the API JSON response to stdout. Non-2xx responses exit with a concise message.
# - Tip: pipe output to a file: perl ... > response.json

binmode STDOUT, ':raw';
binmode STDERR, ':encoding(UTF-8)';

# Load .env from the Perl folder root
my $env_path = "$Bin/../../.env";
-e $env_path and Dotenv->load($env_path);

my $api_key = $ENV{PDFREST_API_KEY} // '';
if (!$api_key || $api_key =~ /^\s*$/) {
    print STDERR "Missing PDFREST_API_KEY in .env or environment\n";
    exit 1;
}

my $api_base = $ENV{PDFREST_URL} // $ENV{PDFREST_API} // 'https://api.pdfrest.com';
$api_base =~ s{/+$}{};

my $pdf_path = shift @ARGV;
if (!$pdf_path || !-f $pdf_path) {
    print STDERR "Usage: perl markdown.pl /path/to/file.pdf\n";
    exit 1;
}

my $filename = basename($pdf_path);
my $ua = LWP::UserAgent->new( timeout => 60 );

eval {
    my $req = POST("$api_base/markdown",
        'Content_Type' => 'form-data',
        'Content' => [
            file => [$pdf_path, $filename, 'Content-Type' => 'application/pdf'],
            output => 'pdfrest_markdown',
            page_break_comments => 'on',
            # Optional: page_range => '1-3',
        ]
    );
    $req->header('api-key' => $api_key);

    my $resp = $ua->request($req);
    print STDOUT $resp->decoded_content // '';
    if (!$resp->is_success) {
        print STDERR "\nMarkdown conversion failed with status " . $resp->code . "\n";
        exit 1;
    }
    1;
} or do {
    my $err = $@ || 'Unknown error';
    $err =~ s/\s+$//;
    print STDERR "Error: $err\n";
    exit 1;
};

__END__
